mount/mount_unix.go

```diff
@@ -37,35 +39,56 @@ func Unmount(target string) error {
 	}
 }
 
-// RecursiveUnmount unmounts the target and all mounts underneath, starting
-// with the deepest mount first. The argument does not have to be a mount
-// point itself.
-func RecursiveUnmount(target string) error {
-	// Fast path, works if target is a mount point that can be unmounted.
+// UnmountAll unmounts all mounts and submounts underneath parent,
+func UnmountAll(parent string) error {
+	// Get all mounts in "parent"
+	mounts, err := mountinfo.GetMounts(mountinfo.PrefixFilter(parent))
+	if err != nil {
+		return err
+	}
+
+	// Fast path: try to unmount top-level mounts first. This works if target is
+	// a mount point that can be unmounted.
 	// On Linux, mntDetach flag ensures a recursive unmount.  For other
 	// platforms, if there are submounts, we'll get EBUSY (and fall back
-	// to the slow path). NOTE we do not ignore EINVAL here as target might
-	// not be a mount point itself (but there can be mounts underneath).
-	if err := unix.Unmount(target, mntDetach); err == nil {
-		return nil
+	// to the slow path). We're not using RecursiveUnmount() here, to avoid
+	// repeatedly calling mountinfo.GetMounts()
+
+	var skipParents []string
+	for _, m := range mounts {
+		// Skip parent itself, and skip non-top-level mounts
+		if m.Mountpoint == parent || path.Dir(m.Mountpoint) != parent {
```

Comment:
So, I think this is not entirely correct optimization. Let me describe what I think would be a correct one.

We have a list of mounts, it's a tree, and we assume that if we use `mntDetach` flag we will only need to unmount the top ones. But the top ones are not those whose parent is `parent`, but those who have submounts.

For example, if we have the following directory tree (mounts are marked with `[n]`):

```
.
├── [1]
│   ├── [4]
│   └── [5]
├── 2
│   └── [6]
└── 3
    └── 7
        └── [8]
```

the top-level mounts are [1], [6], and [8], and we should try unmounting them first, hoping that [4] and [5] will be unmounted.

One way of doing that is to sort the tree so the shortest paths will be first, and upon successful unmount remove the children of the unmounted directory.

I am not that great at algorithms and not sure if it's possible to create the one described above, which will be faster that just iterating over a list and unmounting everything. Maybe it will be possible and even efficient if we use a real tree-like structure and do a width-first tree traversal, discarding the entire branch underneath if umount is successful.

My gut feeling though is that the "smart" approach will be slower than the "stupid" one.

Reply:
One other things is, for special cases, when we know that the top mounts are right below the parent (which is what your code does in fast path), we can avoid reading mountinfo and just do something like
```go
entries, err := os.Readdirnames(parent)
if err != nil {
   return err
}
for _, de := range entries {
   mp := filepath.Join(parent, de)
   err := unix.Unmount(mp, mntDetach)
   if err != nil && err != unix.EINVAL {
      return &os.PathError{Op: "UnmountAll", Path: mp, Err: err}
   }
}
```

after which we can read mountinfo and see if there's anything left.

Now, again, this optimization only makes sense if we assume that in most of the cases the mounts will be a direct descendants of the `parent` (which is also true for your code).